src/mito2/src/memtable/time_series.rs

```diff
@@ -0,0 +1,840 @@
+        );
+        let config = config.unwrap_or_default();
+        config.validate()?;
+        let series_set = SeriesSet::new(
+            region_metadata.clone(),
+            config.bucket_num as u8,
+            config.initial_builder_capacity,
+        );
+        Ok(Self {
+            id,
+            config,
+            region_metadata,
+            series_set,
+            row_codec,
+        })
+    }
+}
+
+impl Debug for TimeSeriesMemtable {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TimeSeriesMemtable")
+            .field("config", &self.config)
+            .finish()
+    }
+}
+
+impl Memtable for TimeSeriesMemtable {
+    fn id(&self) -> MemtableId {
+        self.id
+    }
+
+    fn write(&self, kvs: &KeyValues) -> Result<()> {
+        for kv in kvs.iter() {
+            ensure!(
+                kv.num_primary_keys() == self.row_codec.num_fields(),
+                PrimaryKeyLengthMismatchSnafu {
+                    expect: self.row_codec.num_fields(),
+                    actual: kv.num_primary_keys()
+                }
+            );
+            let primary_key_encoded = self.row_codec.encode(kv.primary_keys())?;
+            let fields = kv.fields().collect();
+            let series = self.series_set.get_or_add_series(primary_key_encoded);
+            let mut guard = series.write().unwrap();
+            guard.push(kv.timestamp(), kv.sequence(), kv.op_type(), fields);
+        }
+        Ok(())
+    }
+
+    fn iter(&self, req: ScanRequest) -> BoxedBatchIterator {
+        let _projection = req.projection.map(|p| {
+            p.iter()
+                .map(|idx| self.region_metadata.column_metadatas[*idx].column_id)
+                .collect::<Vec<_>>()
+        });
+
+        Box::new(self.series_set.iter_series())
+    }
+}
+
+/// Index of series, including bucket index part (u8) and series index inside the bucket (usize).  
+type SeriesIndex = (u8, usize);
+
+struct SeriesSet {
+    initial_builder_capacity: usize,
+    region_metadata: RegionMetadataRef,
+    bucket_num: u8,
+    indices: Arc<RwLock<BTreeMap<Vec<u8>, SeriesIndex>>>,
+    buckets: Arc<Vec<Bucket>>,
+}
+
+impl SeriesSet {
+    fn new(
+        region_metadata: RegionMetadataRef,
+        bucket_num: u8,
+        initial_builder_capacity: usize,
+    ) -> Self {
+        let buckets = Arc::new(
+            (0..bucket_num)
+                .map(|_| Bucket::new(region_metadata.clone()))
+                .collect(),
+        );
+
+        Self {
+            initial_builder_capacity,
+            region_metadata,
+            bucket_num,
+            indices: Default::default(),
+            buckets,
+        }
+    }
+}
+
+impl SeriesSet {
+    /// Returns [Series] at given bucket and index.
+    fn get_bucket_by_index(&self, idx: &SeriesIndex) -> Option<Arc<RwLock<Series>>> {
+        debug_assert!((idx.0 as usize) < self.buckets.len());
+        self.buckets[idx.0 as usize].get_series(idx.1)
```

Comment:
`buckets[idx]` already does the assertion.
```suggestion
        self.buckets[idx.0 as usize].get_series(idx.1)
```